ambry-router/src/main/java/com.github.ambry.router/AdaptiveOperationTracker.java

```diff
@@ -0,0 +1,211 @@
+    }
+    this.otIterator = new OpTrackerIterator();
+  }
+
+  @Override
+  public boolean hasSucceeded() {
+    return succeededCount >= successTarget;
+  }
+
+  @Override
+  public boolean isDone() {
+    return hasSucceeded() || hasFailed();
+  }
+
+  @Override
+  public void onResponse(ReplicaId replicaId, boolean isSuccessFul) {
+    inflightCount--;
+    if (isSuccessFul) {
+      succeededCount++;
+    } else {
+      failedCount++;
+    }
+    getTracker(replicaId).update(time.milliseconds() - replicaRequestSendTimes.remove(replicaId));
+  }
+
+  @Override
+  public Iterator<ReplicaId> getReplicaIterator() {
+    replicaIterator = replicaPool.iterator();
+    return otIterator;
+  }
+
+  /**
+   * @return {@code true} if the operation has failed.
+   */
+  private boolean hasFailed() {
+    return (totalReplicaCount - failedCount) < successTarget;
+  }
+
+  /**
+   * Gets the {@link Histogram} that tracks requests to the class of replicas (intra or inter DC) that {@code replicaId}
+   * belongs to.
+   * @param replicaId the {@link ReplicaId} whose request latency is going to be tracked.
+   * @return the {@link Histogram} that tracks requests to the class of replicas (intra or inter DC) that
+   * {@code replicaId} belongs to.
+   */
+  private Histogram getTracker(ReplicaId replicaId) {
+    if (replicaId.getDataNodeId().getDatacenterName().equals(datacenterName)) {
+      return localColoTracker;
+    }
+    return crossColoTracker;
+  }
+
+  /**
+   * An iterator to fetch replicas to send requests to. Respects parallelism and discounts requests that have been
+   * outstanding for a certain amount of time from parallelism.
+   */
+  private class OpTrackerIterator implements Iterator<ReplicaId> {
+
+    @Override
+    public boolean hasNext() {
+      return replicaIterator.hasNext() && (inflightCount < parallelism || isMostRecentRequestPastDue());
+    }
+
+    @Override
+    public void remove() {
+      replicaIterator.remove();
+      replicaRequestSendTimes.put(lastReturned, time.milliseconds());
+      mostRecentlyTracked = lastReturned;
+      inflightCount++;
+    }
+
+    @Override
+    public ReplicaId next() {
+      if (!hasNext()) {
+        throw new NoSuchElementException();
+      }
+      lastReturned = replicaIterator.next();
+      return lastReturned;
+    }
+
+    /**
+     * @return {@code true} if the most recent request that was sent has been outstanding for more than the cutoff
+     * latency or if a response has already been received for it.
+     * @throws IllegalStateException if no requests have been sent yet.
+     */
+    private boolean isMostRecentRequestPastDue() {
+      if (replicaRequestSendTimes.size() == 0) {
+        throw new IllegalStateException("There have been no requests");
+      }
+      if (!replicaRequestSendTimes.containsKey(mostRecentlyTracked)) {
+        // if the most recently tracked request has already had a response, it is considered past due since
```

Comment:
Your point about being more concerned about the older request is correct. Looking for only the most recent is incorrect and is not true to the statement that "any requests over 90th percentile is not considered towards parallelism". I have fixed that (and a test case). Please take a look at the refactoring.